### PR TITLE
Allowing casting str to int in JIT

### LINKED ISSUE #35948: JIT: int(str) fails
## 🐛 Bug

int(some_string) fails when doing torch.jit.script:

## To Reproduce
```
def f(s: str):
    return int(s)
f = torch.jit.script(f)
```
results in error:
```
  File "C:\Miniconda3\lib\site-packages\torch\jit\__init__.py", line 1281, in script
    fn = torch._C._jit_script_compile(qualified_name, ast, _rcb, get_default_args(obj))
RuntimeError: 
Arguments for call are not valid.
The following variants are available:
  
  aten::Int(Tensor a) -> (int):
  Expected a value of type 'Tensor' for argument 'a' but instead found type 'str'.
  
  aten::Int(bool a) -> (int):
  Expected a value of type 'bool' for argument 'a' but instead found type 'str'.
  
  aten::Int(float a) -> (int):
  Expected a value of type 'float' for argument 'a' but instead found type 'str'.
  
  aten::Int(Scalar a) -> (int):
  Expected a value of type 'number' for argument 'a' but instead found type 'str'.

The original call is:
  File "D:/GitHub/yolov3/qwe.py", line 6
def f(s: str):
    return int(s)
           ~~~ <--- HERE
```

 - PyTorch Version (e.g., 1.0): 1.4.0 (current stable)
 - OS (e.g., Linux): Windows 10
 - How you installed PyTorch (`conda`, `pip`, source): pip install torch===1.4.0 torchvision===0.5.0 -f https://download.pytorch.org/whl/torch_stable.html
 - Build command you used (if compiling from source): --
 - Python version: 3.6.10
 - CUDA/cuDNN version: 10.0
 - GPU models and configuration: 1050
 - Any other relevant information:

cc @suo

### PR DESCRIPTION
**Pull request by vishwakftw**
Changelog:
- Allow int(str) in TorchScript

Test Plan:
- Added tests in test_jit.py

Closes #35948

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -7858,6 +7858,17 @@ def to_str(x):
 
         self.assertEqual("(1, 1)", to_str(1))
 
+    def test_int_cast(self):
+        @torch.jit.script
+        def to_int(x):
+            # type: (str) -> int
+            return int(x)
+
+        self.assertEqual(5, to_int('5'))
```
Comment (driazati) [inline test/test_jit.py:7867]: Can you add some more test cases for things like `"5 "` and the failure cases like `"0x20"` (we leave the `base` arg for a follow up PR)

File: torch/csrc/jit/runtime/register_prim_ops_c10.cpp [hunk 2]
```diff
@@ -33,4 +34,17 @@ static auto registry_prim =
...
+                .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA))
+        .op("aten::Int.str(str a) -> int",
+            torch::RegisterOperators::options()
+                .catchAllKernel(
+                    [](std::string str) -> int64_t {
+                        std::string::size_type sz;
+                        int64_t val = static_cast<int64_t>(c10::stoll(str, &sz));
+                        if (sz != str.size()) {
+                            throw std::runtime_error(
+                                "int() only accepts a string of a single integer");
```
Comment (driazati) [inline torch/csrc/jit/runtime/register_prim_ops_c10.cpp:46]: Can you have this match Python more closely? Something like

`invalid literal for int() with base 10: ' + str + "'"`

**Review by driazati — COMMENTED** [review]
Looks good, a couple small things

File: test/test_jit.py (hunk 1 above)
Comment (vishwakftw) [reply test/test_jit.py:7867]: Sure, I have added tests for non-decimal numbers. However, there seems to be a slightly incorrect implementation of `stoll` and `stod` in `c10` which doesn't trim spaces before passing it to the stringstream. This means that even a statement like `float('5.5  ')` throws an error.

File: torch/csrc/jit/runtime/register_prim_ops_c10.cpp (hunk 2 above)
Comment (vishwakftw) [reply torch/csrc/jit/runtime/register_prim_ops_c10.cpp]: Modified the error messages for both `int()` and `float()`.

**Review by driazati — COMMENTED** [review]
Some of the expected test error messages don't match anymore, could you fix those? Then we can land this

**Comment by vishwakftw** [comment]
@driazati I have addressed the test failures, should be good to go now.

**Review by driazati — APPROVED** [review]

**Comment by vishwakftw** [comment]
@driazati is this good to go?

**Pull request closed without merging**